html/semantics/embedded-content/the-iframe-element/src-repeated-in-ancestor.html

```diff
@@ -0,0 +1,129 @@
+<!doctype html>
+<meta charset="utf-8">
+<title>Navigation should not occur when `src` matches the location of a anscestor browsing context</title>
+<script>
+// Avoid recursion in non-conforming browsers
+if (parent !== window && parent.title == window.title) {
+  window.stop();
+}
+</script>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<body>
+<script>
+/**
+ * This test uses the `beforeunload` event to detect navigation. Because that
+ * event is fired synchronously in response to "process the iframe attributes",
+ * a second "control" iframe may be used to verify cases where navigation
+ * should *not* occur. `Promise.race` ensures that tests complete as soon as
+ * possible.
+ */
+'use strict';
+
+function when(target, eventName) {
+  return new Promise(function(resolve, reject) {
+    target.addEventListener(eventName, function handle() {
+      target.removeEventListener(eventName, handle);
+      resolve();
+    });
+    target.addEventListener('error', function() {
+      reject(new Error('Error while waiting for ' + eventName));
+    });
+  });
+}
+
+function init(doc, t) {
+  var iframes = [doc.createElement('iframe'), doc.createElement('iframe')];
+
+  iframes.forEach(function(iframe) {
+    iframe.src = '/common/blank.html';
+    doc.body.appendChild(iframe);
+
+    t.add_cleanup(function() {
+      iframe.parentNode.removeChild(iframe);
+    });
+  });
+
+  return Promise.all([when(iframes[0], 'load'), when(iframes[1], 'load')])
+    .then(function() { return iframes; });
+}
+
+promise_test(function(t) {
+  return init(document, t)
+    .then(function(iframes) {
+      var subjectNavigation = when(iframes[0].contentWindow, 'beforeunload');
+      var controlNavigation = when(iframes[1].contentWindow, 'beforeunload');
+
+      iframes[0].src = '/common/blank.html?2';
+      iframes[1].src = '/common/blank.html?3';
+
+      return Promise.all([subjectNavigation, controlNavigation]);
+    });
+}, 'different path name');
+
+promise_test(function(t) {
+  return init(document, t)
+    .then(function(iframes) {
+      var subjectNavigation = when(iframes[0].contentWindow, 'beforeunload');
+      var controlNavigation = when(iframes[1].contentWindow, 'beforeunload');
+
+      iframes[0].src = location.pathname;
+      iframes[1].src = '/common/blank.html?4';
+
+      return Promise.race([
```

Comment:
So I'm a little confused.  If the beforeunload event fires sync (before the src setter returns, right?) then why does this need the promise bits?

It seems like this setup could be much simpler and just check that beforeunload fires during the src set for the different-url case but not the same-url case...

Reply:
This strategy is a layover from my earlier attempts to hook into the "load" event. A simplification like that would be nice, but Chrome is firing the event asynchronously, and I can't tell if that's allowed or not.

I believe the text that starts this off is:

> whenever an iframe element with a non-null nested browsing context but with no srcdoc attribute specified has its src attribute set, changed, or removed, the user agent must process the iframe attributes.

That text doesn't seem to require synchronous behavior, so my initial claim about synchronicity may have been incorrect. Then again, since it's not in an algorithm, I'm not sure if it's normative. [The description of the `src` attribute](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#attr-iframe-src) doesn't mention this directly, and I'm not sure where else the relevant normative text would go. Do you have any thoughts on this?

Reply:
It's normative because it has the word "must". It doesn't say to queue a task or anything, therefore it's synchronous.